### PR TITLE
Ipv6 registries

### PR DESCRIPTION
**Pull request by aojea**
registry: support ipv6 addresses
    
Current registry reference use a subset of dns and IPv4 addresses to
represent a registry domain.
Since registries are mostly compatible with rfc3986, that defines the
URI generic syntax, this adds support for IPv6 enclosed in squared
brackets based on the mentioned rfc, but excluding zone identifiers
as defined by rfc6874 or special addresses such as IPv4-Mapped)

Ref: https://github.com/kubernetes/kubernetes/issues/104491

This builds on top of previous PRs

- Allow IPv6 address literals in DomainRegexp distribution/distribution#2899
- Add IPv6 support for domain parsing distribution/distribution#2575

Kudos to those persons for showing the path

### CODE REVIEW

**Comment by aojea** [comment]
Since this was reported in Kubernetes, I'd prefer the whole ecosystem to be consistent about registries URIs, rather than fix it  in Kubernetes only.
cc :@dims @mikebrow @thaJeztah

**Comment by thaJeztah** [comment]
Also wondering if we can use `net.SplitHostPort()` for these 🤔 https://pkg.go.dev/net#SplitHostPort

**Comment by aojea** [comment]
> Also wondering if we can use `net.SplitHostPort()` for these  https://pkg.go.dev/net#SplitHostPort

I used the regex based on current state and previous PRs, but I rather prefer to validate with the golang stdlib functions if possible, it seems less error prone (unless stdlib changes of course, but this only happened in 1.17 for ipv4 addresses)

**Comment by thaJeztah** [comment]
Yes, perhaps there's other reasons for the regexes that I didn't think of, but I saw them, and know that it's easy to make a mistake in regexes (besides them known to be potentially more power-hungry), so _if possible_, using stdlib, that would be my preference.

Of course that's orthogonal to the _" the whole ecosystem to be consistent about registries URIs, rather than fix it in Kubernetes only."_ question

For that, given that currently, `IPv4[:<port>]` is accepted, I (personally) see no reason why the same shouldn't be supported for IPv6. Assuming, of course it doesn't conflict with existing expectations, but (first glance) this PR is only adding _more_ test-cases, and all existing tests continue to pass, so it looks solid.

**Comment by milosgajdos** [comment]
can you please rebase @aojea

**Comment by aojea** [comment]
> can you please rebase @aojea

@milosgajdos rebased :)

**Review by milosgajdos — APPROVED** [review]

**Comment by aojea** [comment]
it needs other approval, right?

**Review by mikebrow — COMMENTED** [review]
see comments

File: reference/reference.go
Comment (aojea) [reply reference/reference.go:13]: sounds good

Comment (aojea) [reply reference/reference.go:13]: new PR for the ports https://github.com/distribution/distribution/pull/3667

I appreciate if we keep the order, and the ports PR goes first, so I don't have to fix the git conflicts :smile:

File: reference/regexp.go
Comment (aojea) [reply reference/regexp.go:31]: I've done some archeology, and seems this is the source of this regex 
https://stackoverflow.com/a/17871737
https://gist.github.com/syzdek/6086792

File: reference/reference.go
Comment (aojea) [reply reference/reference.go:12]: I've added the rfc comments in the other file, that seems cohesive with current status

**Comment by thaJeztah** [comment]
Let me post my comment form https://github.com/distribution/distribution/pull/3667 here as well

> my concern in that one specifically (but partially this one as well), is the huge dependency on more-and-more complex regexes; I know there's been efforts to simplify them (as they are notorious for being a resource-hog).
> 
> From that perspective (but I'd have to have a closer look at how they're all used), I'd rather have a "good enough" regex to split the elements, followed by a "regular" check to make further determination, e.g.;
> 
> 1. split first element if it contains a `.`, a `[` or a `:` (in which case we "assume" it's a hostname or IP-address)
> 2. call `net.SplitHostPort` (and/or other functions) to perform more specific validation
> 
> With that, we avoid running a (likely) heavy regex for each use _and_ will be able to return a more specific error message ("port out of range"), instead of "invalid reference".

**Comment by aojea** [comment]
/hold

let me work in the stdlib approach, I do prefer that too, this IPv6 regex is not something easier to think about

**Comment by aojea** [comment]
ok, it seems I had a misconception, we don't did full validation, as @thaJeztah says we just need a simpler regex 

I was comparing with golang url.Parse and it doesn't perform validation too
https://go.dev/play/p/D0m9NZ1AMJL

**Comment by aojea** [comment]
I think I'm close now, just missing the referenceRegex

**Comment by aojea** [comment]
> I think I'm close now, just missing the referenceRegex

Ok, this is ready

Ipv4 has dots and digits, that are valid repository names, ipv6 has invalid character for repository names.
However both ipv6 and ipv4.will be valid domains

**Comment by aojea** [comment]
@thaJeztah @mikebrow @milosgajdos 

the current approach is consistent with the golang url.Parse implementation, it just parses without validation,

**Review by thaJeztah — APPROVED** [review]
LGTM

we should look at changing some of these variables to `const`'s and see if we need all those utility functions (in favor of just plain string concatenating)

**Comment by aojea** [comment]
@milosgajdos I'm not familiar with the workflow here, what is missing?

**Comment by aojea** [comment]
> @aojea sorry for the radio silence on my end. LGTM.

no worries,

thanks so much

**Comment by aojea** [comment]
by the way, how often do you use to release, so I can revendor it in kubernetes to fix https://github.com/kubernetes/kubernetes/issues/104491

**Comment by milosgajdos** [comment]
We don't have a regular cadence. `main` is currently set up to become `v3` but for that we'll have to have a proper release plan which is not quite the case at the moment. I shall try and put something together, though I wouldn't expect that to become the reality earlier than the EOY 😞

**Comment by aojea** [comment]
Fair enough

**Pull request merged**
